File: tests/cpp/test_metadata.cc
/*
 * Licensed to the Apache Software Foundation (ASF) under one
 * or more contributor license agreements.  See the NOTICE file
 * distributed with this work for additional information
 * regarding copyright ownership.  The ASF licenses this file
 * to you under the Apache License, Version 2.0 (the
 * "License"); you may not use this file except in compliance
 * with the License.  You may obtain a copy of the License at
 *
 *   http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing,
 * software distributed under the License is distributed on an
 * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
 * KIND, either express or implied.  See the License for the
 * specific language governing permissions and limitations
 * under the License.
 */
#include <gtest/gtest.h>
#include <tvm/ffi/c_api.h>
#include <tvm/ffi/container/array.h>
#include <tvm/ffi/container/map.h>
#include <tvm/ffi/container/variant.h>
#include <tvm/ffi/dtype.h>
#include <tvm/ffi/extra/json.h>
#include <tvm/ffi/function.h>
#include <tvm/ffi/object.h>
#include <tvm/ffi/optional.h>
#include <tvm/ffi/reflection/accessor.h>
#include <tvm/ffi/string.h>

namespace {

using namespace tvm::ffi;
using namespace tvm::ffi::reflection;

static std::string ParseMetadataToSchema(const String& metadata) {
  return json::Parse(metadata)
      .cast<Map<String, Any>>()["type_schema"]  //
      .cast<String>();                          //
}

static std::string ParseMetadataToSchema(const TVMFFIByteArray& metadata) {
  return json::Parse(String(metadata))
      .cast<Map<String, Any>>()["type_schema"]  //
      .cast<String>();                          //
}

TEST(Schema, GlobalFuncTypeSchema) {
  // Helper to fetch global function type schema via the exposed utility
  Function get_metadata = Function::GetGlobalRequired("ffi.GetGlobalFuncMetadata");
  auto fetch = [&](const char* name) -> std::string {
    String metadata = get_metadata(String(name)).cast<String>();
    return ParseMetadataToSchema(metadata);
  };
  // Simple IDs
  EXPECT_EQ(fetch("testing.schema_id_int"),
            R"({"type":"ffi.Function","args":[{"type":"int"},{"type":"int"}]})");
  EXPECT_EQ(fetch("testing.schema_id_float"),
            R"({"type":"ffi.Function","args":[{"type":"float"},{"type":"float"}]})");
  EXPECT_EQ(fetch("testing.schema_id_bool"),
            R"({"type":"ffi.Function","args":[{"type":"bool"},{"type":"bool"}]})");
  EXPECT_EQ(fetch("testing.schema_id_device"),
            R"({"type":"ffi.Function","args":[{"type":"Device"},{"type":"Device"}]})");
  EXPECT_EQ(fetch("testing.schema_id_dtype"),
            R"({"type":"ffi.Function","args":[{"type":"DataType"},{"type":"DataType"}]})");
  EXPECT_EQ(fetch("testing.schema_id_string"),
            R"({"type":"ffi.Function","args":[{"type":"ffi.String"},{"type":"ffi.String"}]})");
  EXPECT_EQ(fetch("testing.schema_id_bytes"),
            R"({"type":"ffi.Function","args":[{"type":"ffi.Bytes"},{"type":"ffi.Bytes"}]})");
  EXPECT_EQ(fetch("testing.schema_id_func"),
            R"({"type":"ffi.Function","args":[{"type":"ffi.Function"},{"type":"ffi.Function"}]})");
  EXPECT_EQ(
      fetch("testing.schema_id_func_typed"),
      R"({"type":"ffi.Function","args":[{"type":"ffi.Function","args":[{"type":"None"},{"type":"int"},{"type":"float"},{"type":"ffi.Function"}]},{"type":"ffi.Function","args":[{"type":"None"},{"type":"int"},{"type":"float"},{"type":"ffi.Function"}]}]})");

  EXPECT_EQ(fetch("testing.schema_id_any"),
            R"({"type":"ffi.Function","args":[{"type":"Any"},{"type":"Any"}]})");
  EXPECT_EQ(fetch("testing.schema_id_object"),
            R"({"type":"ffi.Function","args":[{"type":"ffi.Object"},{"type":"ffi.Object"}]})");
  EXPECT_EQ(fetch("testing.schema_id_dltensor"),
            R"({"type":"ffi.Function","args":[{"type":"DLTensor*"},{"type":"DLTensor*"}]})");
  EXPECT_EQ(fetch("testing.schema_id_tensor"),
            R"({"type":"ffi.Function","args":[{"type":"ffi.Tensor"},{"type":"ffi.Tensor"}]})");
  EXPECT_EQ(fetch("testing.schema_tensor_view_input"),
            R"({"type":"ffi.Function","args":[{"type":"None"},{"type":"DLTensor*"}]})");
  EXPECT_EQ(
      fetch("testing.schema_id_opt_int"),
      R"({"type":"ffi.Function","args":[{"type":"Optional","args":[{"type":"int"}]},{"type":"Optional","args":[{"type":"int"}]}]})");
  EXPECT_EQ(
      fetch("testing.schema_id_opt_str"),
      R"({"type":"ffi.Function","args":[{"type":"Optional","args":[{"type":"ffi.String"}]},{"type":"Optional","args":[{"type":"ffi.String"}]}]})");
  EXPECT_EQ(
      fetch("testing.schema_id_opt_obj"),
      R"({"type":"ffi.Function","args":[{"type":"Optional","args":[{"type":"ffi.Object"}]},{"type":"Optional","args":[{"type":"ffi.Object"}]}]})");
  EXPECT_EQ(
      fetch("testing.schema_id_arr_int"),
      R"({"type":"ffi.Function","args":[{"type":"ffi.Array","args":[{"type":"int"}]},{"type":"ffi.Array","args":[{"type":"int"}]}]})");
  EXPECT_EQ(
      fetch("testing.schema_id_arr_str"),
      R"({"type":"ffi.Function","args":[{"type":"ffi.Array","args":[{"type":"ffi.String"}]},{"type":"ffi.Array","args":[{"type":"ffi.String"}]}]})");
  EXPECT_EQ(
      fetch("testing.schema_id_arr_obj"),
      R"({"type":"ffi.Function","args":[{"type":"ffi.Array","args":[{"type":"ffi.Object"}]},{"type":"ffi.Array","args":[{"type":"ffi.Object"}]}]})");
  EXPECT_EQ(fetch("testing.schema_id_arr"),
            R"({"type":"ffi.Function","args":[{"type":"ffi.Array"},{"type":"ffi.Array"}]})");
  EXPECT_EQ(
      fetch("testing.schema_id_map_str_int"),
      R"({"type":"ffi.Function","args":[{"type":"ffi.Map","args":[{"type":"ffi.String"},{"type":"int"}]},{"type":"ffi.Map","args":[{"type":"ffi.String"},{"type":"int"}]}]})");
  EXPECT_EQ(
      fetch("testing.schema_id_map_str_str"),
      R"({"type":"ffi.Function","args":[{"type":"ffi.Map","args":[{"type":"ffi.String"},{"type":"ffi.String"}]},{"type":"ffi.Map","args":[{"type":"ffi.String"},{"type":"ffi.String"}]}]})");
  EXPECT_EQ(
      fetch("testing.schema_id_map_str_obj"),
      R"({"type":"ffi.Function","args":[{"type":"ffi.Map","args":[{"type":"ffi.String"},{"type":"ffi.Object"}]},{"type":"ffi.Map","args":[{"type":"ffi.String"},{"type":"ffi.Object"}]}]})");
  EXPECT_EQ(fetch("testing.schema_id_map"),
            R"({"type":"ffi.Function","args":[{"type":"ffi.Map"},{"type":"ffi.Map"}]})");
  EXPECT_EQ(
      fetch("testing.schema_id_variant_int_str"),
      R"({"type":"ffi.Function","args":[{"type":"Variant","args":[{"type":"int"},{"type":"ffi.String"}]},{"type":"Variant","args":[{"type":"int"},{"type":"ffi.String"}]}]})");

  // Packed function registered via def_packed: schema is plain ffi.Function
  EXPECT_EQ(fetch("testing.schema_packed"), R"({"type":"ffi.Function"})");

  // Mixed containers and optionals
  EXPECT_EQ(
      fetch("testing.schema_arr_map_opt"),
      R"({"type":"ffi.Function","args":[{"type":"ffi.Map","args":[{"type":"ffi.String"},{"type":"ffi.Array","args":[{"type":"int"}]}]},{"type":"ffi.Array","args":[{"type":"Optional","args":[{"type":"int"}]}]},{"type":"ffi.Map","args":[{"type":"ffi.String"},{"type":"ffi.Array","args":[{"type":"int"}]}]},{"type":"Optional","args":[{"type":"ffi.String"}]}]})");

  EXPECT_EQ(
      fetch("testing.schema_variant_mix"),
      R"({"type":"ffi.Function","args":[{"type":"Variant","args":[{"type":"int"},{"type":"ffi.String"},{"type":"ffi.Array","args":[{"type":"int"}]}]},{"type":"Variant","args":[{"type":"int"},{"type":"ffi.String"},{"type":"ffi.Array","args":[{"type":"int"}]}]}]})");

  // No-arg and no-return combinations
  EXPECT_EQ(fetch("testing.schema_no_args"), R"({"type":"ffi.Function","args":[{"type":"int"}]})");
  EXPECT_EQ(fetch("testing.schema_no_return"),
            R"({"type":"ffi.Function","args":[{"type":"None"},{"type":"int"}]})");
  EXPECT_EQ(fetch("testing.schema_no_args_no_return"),
            R"({"type":"ffi.Function","args":[{"type":"None"}]})");
}

TEST(Schema, FieldTypeSchemas) {
  // Validate type schema JSON on fields of testing.SchemaAllTypes
  const char* kTypeKey = "testing.SchemaAllTypes";
  // Helper to fetch a field's type schema by name
  auto field_schema = [&](const char* field_name) -> std::string {
    const TVMFFIFieldInfo* info = GetFieldInfo(kTypeKey, field_name);
    return ParseMetadataToSchema(info->metadata);
  };

  EXPECT_EQ(field_schema("v_bool"), R"({"type":"bool"})");
  EXPECT_EQ(field_schema("v_int"), R"({"type":"int"})");
  EXPECT_EQ(field_schema("v_float"), R"({"type":"float"})");
  EXPECT_EQ(field_schema("v_device"), R"({"type":"Device"})");
  EXPECT_EQ(field_schema("v_dtype"), R"({"type":"DataType"})");
  EXPECT_EQ(field_schema("v_string"), R"({"type":"ffi.String"})");
  EXPECT_EQ(field_schema("v_bytes"), R"({"type":"ffi.Bytes"})");
  EXPECT_EQ(field_schema("v_opt_int"), R"({"type":"Optional","args":[{"type":"int"}]})");
  EXPECT_EQ(field_schema("v_opt_str"), R"({"type":"Optional","args":[{"type":"ffi.String"}]})");
  EXPECT_EQ(field_schema("v_arr_int"), R"({"type":"ffi.Array","args":[{"type":"int"}]})");
  EXPECT_EQ(field_schema("v_arr_str"), R"({"type":"ffi.Array","args":[{"type":"ffi.String"}]})");
  EXPECT_EQ(field_schema("v_map_str_int"),
            R"({"type":"ffi.Map","args":[{"type":"ffi.String"},{"type":"int"}]})");
  EXPECT_EQ(
      field_schema("v_map_str_arr_int"),
      R"({"type":"ffi.Map","args":[{"type":"ffi.String"},{"type":"ffi.Array","args":[{"type":"int"}]}]})");
  EXPECT_EQ(
      field_schema("v_variant"),
      R"({"type":"Variant","args":[{"type":"ffi.String"},{"type":"ffi.Array","args":[{"type":"int"}]},{"type":"ffi.Map","args":[{"type":"ffi.String"},{"type":"int"}]}]})");
  EXPECT_EQ(
      field_schema("v_opt_arr_variant"),
      R"({"type":"Optional","args":[{"type":"ffi.Array","args":[{"type":"Variant","args":[{"type":"int"},{"type":"ffi.String"}]}]}]})");
}

TEST(Schema, MethodTypeSchemas) {
  const char* kTypeKey = "testing.SchemaAllTypes";
  auto method_schema = [&](const char* method_name) -> std::string {
    const TVMFFIMethodInfo* info = GetMethodInfo(kTypeKey, method_name);
    return ParseMetadataToSchema(info->metadata);
  };

  // Instance methods
  EXPECT_EQ(
      method_schema("add_int"),
      R"({"type":"ffi.Function","args":[{"type":"int"},{"type":"testing.SchemaAllTypes"},{"type":"int"}]})");
  EXPECT_EQ(
      method_schema("append_int"),
      R"({"type":"ffi.Function","args":[{"type":"ffi.Array","args":[{"type":"int"}]},{"type":"testing.SchemaAllTypes"},{"type":"ffi.Array","args":[{"type":"int"}]},{"type":"int"}]})");
  EXPECT_EQ(
      method_schema("maybe_concat"),
      R"({"type":"ffi.Function","args":[{"type":"Optional","args":[{"type":"ffi.String"}]},{"type":"testing.SchemaAllTypes"},{"type":"Optional","args":[{"type":"ffi.String"}]},{"type":"Optional","args":[{"type":"ffi.String"}]}]})");
  EXPECT_EQ(
      method_schema("merge_map"),
      R"({"type":"ffi.Function","args":[{"type":"ffi.Map","args":[{"type":"ffi.String"},{"type":"ffi.Array","args":[{"type":"int"}]}]},{"type":"testing.SchemaAllTypes"},{"type":"ffi.Map","args":[{"type":"ffi.String"},{"type":"ffi.Array","args":[{"type":"int"}]}]},{"type":"ffi.Map","args":[{"type":"ffi.String"},{"type":"ffi.Array","args":[{"type":"int"}]}]}]})");

  // Static method make_with: return type is the object type itself.
  // Build expected JSON as ffi.Function with return type = type_key and args = (int, float, str)
  EXPECT_EQ(
      method_schema("make_with"),
      R"({"type":"ffi.Function","args":[{"type":"testing.SchemaAllTypes"},{"type":"int"},{"type":"float"},{"type":"ffi.String"}]})");
}

}  // namespace

File: tests/python/test_metadata.py
# Licensed to the Apache Software Foundation (ASF) under one
# or more contributor license agreements.  See the NOTICE file
# distributed with this work for additional information
# regarding copyright ownership.  The ASF licenses this file
# to you under the Apache License, Version 2.0 (the
# "License"); you may not use this file except in compliance
# with the License.  You may obtain a copy of the License at
#
#   http://www.apache.org/licenses/LICENSE-2.0
#
# Unless required by applicable law or agreed to in writing,
# software distributed under the License is distributed on an
# "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
# KIND, either express or implied.  See the License for the
# specific language governing permissions and limitations
# under the License.
from typing import Any

import pytest
from tvm_ffi import get_global_func_metadata
from tvm_ffi.core import TypeInfo, TypeSchema
from tvm_ffi.testing import _SchemaAllTypes


def _replace_list_dict(ty: str) -> str:
    return {
        "list": "Sequence",
        "dict": "Mapping",
    }.get(ty, ty)


@pytest.mark.parametrize(
    "func_name,expected",
    [
        ("testing.schema_id_int", "Callable[[int], int]"),
        ("testing.schema_id_float", "Callable[[float], float]"),
        ("testing.schema_id_bool", "Callable[[bool], bool]"),
        ("testing.schema_id_device", "Callable[[Device], Device]"),
        ("testing.schema_id_dtype", "Callable[[dtype], dtype]"),
        ("testing.schema_id_string", "Callable[[str], str]"),
        ("testing.schema_id_bytes", "Callable[[bytes], bytes]"),
        ("testing.schema_id_func", "Callable[[Callable[..., Any]], Callable[..., Any]]"),
        (
            "testing.schema_id_func_typed",
            "Callable[[Callable[[int, float, Callable[..., Any]], None]], Callable[[int, float, Callable[..., Any]], None]]",
        ),
        ("testing.schema_id_any", "Callable[[Any], Any]"),
        ("testing.schema_id_object", "Callable[[Object], Object]"),
        ("testing.schema_id_dltensor", "Callable[[Tensor], Tensor]"),
        ("testing.schema_id_tensor", "Callable[[Tensor], Tensor]"),
        ("testing.schema_tensor_view_input", "Callable[[Tensor], None]"),
        ("testing.schema_id_opt_int", "Callable[[int | None], int | None]"),
        ("testing.schema_id_opt_str", "Callable[[str | None], str | None]"),
        ("testing.schema_id_opt_obj", "Callable[[Object | None], Object | None]"),
        ("testing.schema_id_arr_int", "Callable[[list[int]], list[int]]"),
        ("testing.schema_id_arr_str", "Callable[[list[str]], list[str]]"),
        ("testing.schema_id_arr_obj", "Callable[[list[Object]], list[Object]]"),
        ("testing.schema_id_arr", "Callable[[list[Any]], list[Any]]"),
        ("testing.schema_id_map_str_int", "Callable[[dict[str, int]], dict[str, int]]"),
        ("testing.schema_id_map_str_str", "Callable[[dict[str, str]], dict[str, str]]"),
        ("testing.schema_id_map_str_obj", "Callable[[dict[str, Object]], dict[str, Object]]"),
        ("testing.schema_id_map", "Callable[[dict[Any, Any]], dict[Any, Any]]"),
        ("testing.schema_id_variant_int_str", "Callable[[int | str], int | str]"),
        ("testing.schema_packed", "Callable[..., Any]"),
        (
            "testing.schema_arr_map_opt",
            "Callable[[list[int | None], dict[str, list[int]], str | None], dict[str, list[int]]]",
        ),
        ("testing.schema_variant_mix", "Callable[[int | str | list[int]], int | str | list[int]]"),
        ("testing.schema_no_args", "Callable[[], int]"),
        ("testing.schema_no_return", "Callable[[int], None]"),
        ("testing.schema_no_args_no_return", "Callable[[], None]"),
    ],
)
def test_schema_global_func(func_name: str, expected: str) -> None:
    metadata: dict[str, Any] = get_global_func_metadata(func_name)
    actual: TypeSchema = TypeSchema.from_json_str(metadata["type_schema"])
    assert str(actual) == expected, f"{func_name}: {actual}"
    assert actual.repr(_replace_list_dict) == expected.replace(
        "list",
        "Sequence",
    ).replace(
        "dict",
        "Mapping",
    )


@pytest.mark.parametrize(
    "field_name,expected",
    [
        ("v_bool", "bool"),
        ("v_int", "int"),
        ("v_float", "float"),
        ("v_device", "Device"),
        ("v_dtype", "dtype"),
        ("v_string", "str"),
        ("v_bytes", "bytes"),
        ("v_opt_int", "int | None"),
        ("v_opt_str", "str | None"),
        ("v_arr_int", "list[int]"),
        ("v_arr_str", "list[str]"),
        ("v_map_str_int", "dict[str, int]"),
        ("v_map_str_arr_int", "dict[str, list[int]]"),
        ("v_variant", "str | list[int] | dict[str, int]"),
        ("v_opt_arr_variant", "list[int | str] | None"),
    ],
)
def test_schema_field(field_name: str, expected: str) -> None:
    type_info: TypeInfo = getattr(_SchemaAllTypes, "__tvm_ffi_type_info__")
    for field in type_info.fields:
        if field.name == field_name:
            actual: TypeSchema = TypeSchema.from_json_str(field.metadata["type_schema"])
            assert str(actual) == expected, f"{field_name}: {actual}"
            assert actual.repr(_replace_list_dict) == expected.replace(
                "list",
                "Sequence",
            ).replace(
                "dict",
                "Mapping",
            )
            break
    else:
        raise ValueError(f"Field not found: {field_name}")


@pytest.mark.parametrize(
    "method_name,expected",
    [
        ("add_int", "Callable[[testing.SchemaAllTypes, int], int]"),
        ("append_int", "Callable[[testing.SchemaAllTypes, list[int], int], list[int]]"),
        ("maybe_concat", "Callable[[testing.SchemaAllTypes, str | None, str | None], str | None]"),
        (
            "merge_map",
            "Callable[[testing.SchemaAllTypes, dict[str, list[int]], dict[str, list[int]]], dict[str, list[int]]]",
        ),
        ("make_with", "Callable[[int, float, str], testing.SchemaAllTypes]"),
    ],
)
def test_schema_member_method(method_name: str, expected: str) -> None:
    type_info: TypeInfo = getattr(_SchemaAllTypes, "__tvm_ffi_type_info__")
    for method in type_info.methods:
        if method.name == method_name:
            actual: TypeSchema = TypeSchema.from_json_str(method.metadata["type_schema"])
            assert str(actual) == expected, f"{method_name}: {actual}"
            assert actual.repr(_replace_list_dict) == expected.replace(
                "list",
                "Sequence",
            ).replace(
                "dict",
                "Mapping",
            )
            break
    else:
        raise ValueError(f"Method not found: {method_name}")


def test_metadata_global_func() -> None:
    metadata: dict[str, Any] = get_global_func_metadata("testing.schema_id_int")
    assert len(metadata) == 4
    assert "type_schema" in metadata
    assert metadata["bool_attr"] is True
    assert metadata["int_attr"] == 1
    assert metadata["str_attr"] == "hello"


def test_metadata_field() -> None:
    type_info: TypeInfo = getattr(_SchemaAllTypes, "__tvm_ffi_type_info__")
    for field in type_info.fields:
        if field.name == "v_bool":
            assert len(field.metadata) == 4
            assert "type_schema" in field.metadata
            assert field.metadata["bool_attr"] is True
            assert field.metadata["int_attr"] == 1
            assert field.metadata["str_attr"] == "hello"
            break
    else:
        raise ValueError("Field not found: v_bool")


def test_metadata_member_method() -> None:
    type_info: TypeInfo = getattr(_SchemaAllTypes, "__tvm_ffi_type_info__")
    for method in type_info.methods:
        if method.name == "add_int":
            assert len(method.metadata) == 4
            assert "type_schema" in method.metadata
            assert method.metadata["bool_attr"] is True
            assert method.metadata["int_attr"] == 1
            assert method.metadata["str_attr"] == "hello"
            break
    else:
        raise ValueError("Method not found: add_int")


def test_mem_fn_as_global_func() -> None:
    metadata: dict[str, Any] = get_global_func_metadata("testing.TestIntPairSum")
    type_schema: TypeSchema = TypeSchema.from_json_str(metadata["type_schema"])
    assert str(type_schema) == "Callable[[testing.TestIntPair], int]"
